### DISCUSSION TITLE
Real-time collaboration

### DISCUSSION THREAD

**Discussion by sigaloid** [TriliumNext/Trilium#1634] [Ideas] [answered]
This would be a cool idea to add that would put this on the level of Atom as a code editor. It sounds like a lot of work of course, so if it's out of scope, forgive my rambling. Possibly look into https://github.com/atom/teletype-client (js) as Atom has it implemented pretty well. I'd be glad to host a server that handles the connections on my VPS. 

Some thoughts: 
Websockets? it should be fast enough to handle real time collaboration + I have a little experience with it
Encryption? Never trust the server, but how would peer connections be encrypted? I think a good idea would be an X25519 key however these are probably too long to read off easily, maybe a short PIN or a short 3 word passphrase like wormhole does.
How would Trilium handle peer text input? A simple idea would be just adding an additional module that transcribes the peer's content / keystrokes onto the text, just like the keyboard does. 

In general real time text collaboration networking is not easy and for all I know, I'm the only one who would use this. Thoughts?

**Comment by zadam** [comment, marked as answer]
Hi, even non-real time collaboration is pretty huge thing, realtime is of course even harder.

There's a FAQ entry on this: https://github.com/zadam/trilium/wiki/FAQ#multi-user-support

**Reply by sigaloid** [reply]
Even in the case of a single note sharing, instead of entire databases? I'd think sharing a single note would be a lot less of an undertaking because user management and permissions aren't an issue. conflict resolution real-time editing are what is left, and I will not deny that those are hard. 
>In multi-user app, our scripting support would be a XSS security hole

Unless the user ran it, would it run automatically? In which case I can understand this.

**Reply by zadam** [reply]
It's questionable how useful collaboration on a single note would be. Especially given that Trilium promotes break up of larger notes into a tree structure through features like "Include note" and images being a first class note objects.

> Unless the user ran it, would it run automatically? In which case I can understand this.

Yes, scripts can be run also automatically (based on various triggers, e.g. viewing of a note).

**Reply by trenta3** [reply]
@zadam May I suggest a ready-made CRDT solution like [SyncedStore](https://syncedstore.org), based on Yjs? This does relieve you to think of conflict resolution and may make everything easier.

**Comment by schmurtzm** [comment]
That's right collaboration is really a big missing feature for Trilium Next.
I see three big steps:
- be able to create users accounts for trilium
- be able to share a note to another Trilium Next user in the tree view (probably "easily" doable if the users are on the same server, harder on different server instances).
- be able to edit it simultaneously



When you are using a server you can block access to a section/note by using password protection, it's already a kind of user management in a way. I mean that it seems that the first steps to allow user management on the same server instance seems technically possible, isn't it ?

After the multi user management there is the sharing of note feature, to allow other users to see your shared notes.

And the last step is the real time editing... For this a collaboration with https://github.com/ether/etherpad-lite could be awesome for Trilium Next 🙂

**Reply by CobriMediaJulien** [reply]
TBH i think it´s too much for a "personal" knowledge base.
